Samples/ios/SampleAppiOS-Swift/SampleAppiOS-Swift/SamplePhotoUtil.swift

```diff
@@ -0,0 +1,235 @@
+        
+        if let cachedImagePath = cachedImagePath(), let cachedImage = UIImage(contentsOfFile: cachedImagePath) {
+            currentUserPhoto = cachedImage
+        }
+        else {
+            currentUserPhoto = UIImage(named: "no_photo")
+        }
+        
+        return currentUserPhoto!
+    }
+    
+    // Checks with the graph for an updated photo, if enough time has passed since the last check
+    func checkUpdatePhoto(withCompletion completion: @escaping PhotoCompletion) {
+        if checkTimestamp() == false {
+            return
+        }
+        
+        getUserPhotoImpl {
+            (image, error) in
+            DispatchQueue.main.async {
+                completion(image, error)
+            }
+        }
+    }
+    
+    // Clears out any cached data for the current user
+    func clearPhotoCache() {
+        UserDefaults.standard.removeObject(forKey: kLastPhotoCheckKey)
+        currentUserPhoto = nil
+        
+        if let _ = SampleMSALUtil.shared.currentUserIdentifier {
+            
+            guard let imagePath = cachedImagePath() else {
+                print("User is not signed in. There is nothing to delete")
+                return
+            }
+            
+            do {
+                try FileManager.default.removeItem(at: URL(string: imagePath)!)
+            }
+            catch let error {
+                print("\(error)")
+            }
+        }
+    }
+}
+
+
+// MARK: Caching
+fileprivate extension SamplePhotoUtil {
+    
+    func cachedImageDirectory() -> String {
+        let directories = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true)
+        return "\(directories[0])/com.microsoft.MSALSampleApp/userphoto"
+    }
+    
+    func cachedImagePath() -> String? {
+        if let currentUserIdentifier = SampleMSALUtil.shared.currentUserIdentifier {
+            return cachedImageDirectory() + "/" + currentUserIdentifier
+        }
+        return nil
+    }
+    
+    func setLastChecked() {
+        UserDefaults.standard.set(Data(), forKey: kLastPhotoCheckKey)
+    }
+    
+    func cache(photo data: Data) throws {
+        let imageDirectory = cachedImageDirectory()
+        
+        do {
+            if (FileManager.default.fileExists(atPath: imageDirectory) == false) {
+                try FileManager.default.createDirectory(atPath: imageDirectory, withIntermediateDirectories: true, attributes: [:])
+            }
+            
+            guard let imagePath = cachedImagePath() else {
+                throw SampleAppError.NoUserSignedIn
+            }
+            
+            try data.write(to: URL(fileURLWithPath: imagePath))
+        } catch let error {
+            throw SampleAppError.ImageCacheError(error)
+        }
+    }
+    
+    func checkTimestamp() -> Bool {
+        guard let cachedImagePath = cachedImagePath(),
+            let lastChecked = UserDefaults.standard.object(forKey: kLastPhotoCheckKey) as? Date else {
```

Comment:
nit: this guard statement is a little bit long and difficult to read, maybe worth considering splitting it unless you see benefits of having it like that